### PR TITLE
feat(controller)!: Key-only artifacts. Fixes #3184

### LINKED ISSUE #3184: Define `artifactRepositoryRef` only once in spec
At the moment each new workflow must redefine an artifact repository:

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: input-from-s3-

spec:
  entrypoint: input-from-s3
  serviceAccountName: argo
  templates:
  - name: input-from-s3
    container:
      image: alpine:3.12
      command: [sh, -c]
      args: ["cat /inputs/inputs.txt"]
    inputs:
      artifacts:
        - name: test-input
          path: /inputs
          s3:
            endpoint: s3.amazonaws.com
            region: us-east-2
            bucket: test-bucket
            key: inputs
            accessKeySecret:
              name: argo-s3-credentials
              key: accessKey
            secretKeySecret:
              name: argo-s3-credentials
              key: secretKey
```

Is that possible to define `s3` once and reuse it in different workflows? Something like that:

```yaml
apiVersion: argoproj.io/v1alpha1
kind: ArtifactRepo
metadata:
  name: my-input-artifact
spec:
  type: s3
  config:
    endpoint: s3.amazonaws.com
    region: us-east-1
    bucket: test-hadrien
    key: inputs
    accessKeySecret:
      name: argo-s3-credentials
      key: accessKey
    secretKeySecret:
      name: argo-s3-credentials
      key: secretKey
---
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: input-from-s3-

spec:
  entrypoint: input-from-s3
  serviceAccountName: argo
  templates:
  - name: input-from-s3
    container:
      image: alpine:3.12
      command: [sh, -c]
      args: ["cat /inputs/inputs.txt"]
    inputs:
      artifacts:
        - name: test-input
          path: /inputs
          artifactRepo: my-input-artifact
```

Or maybe this is already possible using workflow templates? If that's the case could you provide an example?

### PR DESCRIPTION
**Pull request by alexec**
Signed-off-by: Alex Collins <alex_collins@intuit.com>

Checklist:

* [x] My organization is added to [USERS.md](https://github.com/argoproj/argo/blob/master/USERS.md).

Fixes #3184

* Allow users to specify artifacts only using the key, bucket and endpoint are taken from the configuration, reducing repetition. 
* Node status only includes the key if needed. Reducing YAML size for large workflows.
* Significant refactoring.
* Lots of new tests.

### CODE REVIEW

**Comment by alexec** [comment]
Closing PR. Will need to rebase after code migration.

**Comment by sarabala1979** [comment]
@alexec Can we add the example and documents?

**Comment by alexec** [comment]
* [ ] add example usage
* [ ] update map-reduce example to use this

**Comment by alexec** [comment]
@sarabala1979 docs and examples added.

File: pkg/apis/workflow/v1alpha1/workflow_types.go [hunk 1]
```diff
@@ -874,70 +868,114 @@ type ArtifactLocation struct {
...
-func (a *ArtifactLocation) HasLocation() bool {
-	return a.S3.HasLocation() ||
-		a.Git.HasLocation() ||
-		a.HTTP.HasLocation() ||
-		a.Artifactory.HasLocation() ||
-		a.Raw.HasLocation() ||
-		a.HDFS.HasLocation() ||
-		a.OSS.HasLocation() ||
-		a.GCS.HasLocation()
+func (a *ArtifactLocation) Get() ArtifactLocationType {
```
Comment (sarabala1979) [inline pkg/apis/workflow/v1alpha1/workflow_types.go:871]: much clear

**Review by sarabala1979 — APPROVED** [review]
LGTM

**Pull request merged**